0_From_Code_First_To_First_Principles/1_code_first/assignment/src/routes/userRoutes.ts

```diff
@@ -0,0 +1,107 @@
+import express from 'express';
+import { areInputsValid } from '../utils';
+import { ERROR_EMAIL_TAKEN, ERROR_USERNAME_TAKEN, ERROR_USER_NOT_FOUND, ERROR_VALIDATION, createUser, editUser, getUserDataByEmail } from '../controllers/userController';
+import { userExists } from '../database';
+
+const userRouter = express.Router();
+
+userRouter.get('/', async (req, res) => {
+  const email = req.query.email;
+
+  if (typeof(email) === 'string' && email.length > 0) {
```

Comment:
Again, great job making it work. Moving forward, we will want to refine this controller to be slim. All we want to see in the solution is that it's minimal.

An improvement we'll need to make here: **putting this validation logic somewhere where it makes more sense.**

Notice that...

```
if (typeof(email) === 'string' && email.length > 0) {
```

... kinda breaks the flow of what it is we're trying to express here? It all of a sudden gets really low level. 

The high-level story of the controller is this: 

_The Controller_

- 1. Validate input
- 2. Pass to use case
- 3. Return data & status code based on response

An analogy I like is that of a book. If this was a book with 3 chapters, I'd want a clear, cohesive table of contents. That's what I expect your code to look like in a way. 

However, the line I've highlighted line here kinda takes us into something that might be more like _Chapter 1.2_ - How to Check If The Email Was Supplied.

Theoretically, the shortest this procedural method this could probably ever be is 3-4 lines, because there are about 3-4 steps to the pseudocode here:

```
const validatedOrInvalidatedData = validateInput(result):

if (validatedOrInvalidatedData.isInvalid()) return RESTfulAPI.createClientError(validatedOrInvalidatedData);

const useCaseResult = await useCase(validatedOrInvalidatedData);

return RESTfulAPI.createResponseFromUseCase(useCaseResult) 
```

That's an overoptimization, and I probably wouldn't express it like that for many design reasons, but just something to be aware of, when you're breaking _cohesion_ and the tradeoffs with fixing it.